app/models/taxon_lineage.rb

```diff
@@ -37,4 +38,57 @@ def self.get_genus_info(genus_tax_id)
     r = find_by(genus_taxid: genus_tax_id)
     return { query: r.genus_name, tax_id: genus_tax_id } if r
   end
+
+  def self.fill_lineage_details(tax_map)
+    # Get list of tax_ids to look up in TaxonLineage and TaxonCount rows. Include family_taxids.
+    tax_ids = tax_map.map { |x| x['tax_id'] }
+    tax_ids |= tax_map.map { |x| x['family_taxid'] }
+
+    # TODO: Should definitely be simplified with taxonomy/lineage refactoring.
+    lineage_by_taxid = {}
+    TaxonLineage.where(taxid: tax_ids).find_each do |x|
+      lineage_by_taxid[x.taxid] = x.as_json
+    end
+
+    # Set lineage info from the first positive tax_id of the species, genus, or family levels.
+    # Preserve names of the negative 'non-specific' nodes.
+    # Used for the tree view and sets appropriate lineage info at each node.
+
+    # Make a new hash with 'species_taxid', 'genus_taxid', etc.
+    missing_vals = Hash[MISSING_LINEAGE_ID.map { |k, v| [k.to_s + "_taxid", v] }]
+
+    tax_map.each do |tax|
+      # Grab the appropriate lineage info by the first positive tax level
+      tax['lineage'] = missing_vals
+      lineage_id = most_specific_positive_id(tax)
+      lin = lineage_by_taxid[lineage_id] if lineage_id
+      tax['lineage'] = lin if lin
```

Comment:
i am unclear on the scope of `lin`, is it a new variable for every iteration of the loop, or is the same variable and the value from one iteration can be seen by the next?

here lin woud be uninitialized if the very first iteration o the loop did not have a positive lineage_id

Reply:
May I recommend
```
tax['lineage'] = lineage_by_taxid[lineage_id] if lineage_id else missing_values.dup
```